proposals/p5087.md

```diff
@@ -0,0 +1,223 @@
+# Qualified lookup into types being defined
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/5087)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Uses requiring a complete type](#uses-requiring-a-complete-type)
+    -   [Lookups into an incomplete generic](#lookups-into-an-incomplete-generic)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Require completeness for qualified name lookup](#require-completeness-for-qualified-name-lookup)
+    -   [Do not require a definition for name lookup](#do-not-require-a-definition-for-name-lookup)
+
+<!-- tocstop -->
+
+## Abstract
+
+Allow qualified name lookup into classes and interfaces as soon as we reach the
+`{` of the definition, rather than disallowing such lookups until we reach the
+`}`.
+
+## Problem
+
+We allow unqualified lookups within a type definition to find names that were
+already declared, but not qualified lookups:
+
+```carbon
+class A {
+  class B {}
+  // ✅ OK, `B` names `A.B`.
+  fn F() -> B;
+  // ❌ Error (before this proposal): `A` is not complete.
+  fn G() -> A.B;
+}
+
+interface I {
+  let T:! type;
+  // ✅ OK, `T` names `Self.T`.
+  fn F() -> T;
+  // ❌ Error (before this proposal): type `I` of `Self` is not complete.
+  fn G() -> Self.T;
+}
+```
+
+This is inconsistent and prevents useful code:
+
+```carbon
+interface Container {
+  let ValueType:! type
+  // ❌ Error (before this proposal): type `Container` of `.Self`
+  // is not complete in implicit access to `.Self.ValueType`.
+  // Implicit access to `Self.ValueType` is OK though.
+  let SliceType:! Container where .ValueType == ValueType;
+}
+```
+
+An additional inconsistency is that namespaces already support qualified lookup
+before the full list of names is known -- indeed, the full list of names in a
+namespace is never known.
+
+## Background
+
+Proposal
+[#3763](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3763.md)
+introduces a name poisoning rule:
+
+> In a declarative scope, it is an error if a name is first looked up and not
+> found, and later introduced.
+
+This is achieved by _poisoning_ a name in a scope when we perform a failed
+lookup for that name in that scope, and diagnosing if the name is later declared
+in a scope where it is poisoned.
+
+With that rule, there is no risk in allowing name lookups into a scope to
+succeed even before the scope is complete. If the name lookup's meaning would be
+changed by a later declaration, an error is issued.
+
```

Comment:
OPTIONAL: could link to the discussion notes on [2025-02-21](https://docs.google.com/document/d/1Iut5f2TQBrtBNIduF4vJYOKfw7MbS8xH_J01_Q4e6Rk/edit?resourcekey=0-mc_vh5UzrzXfU4kO-3tOjA&tab=t.0#heading=h.ix77am1xk6po), [2025-02-27a](https://docs.google.com/document/d/1Iut5f2TQBrtBNIduF4vJYOKfw7MbS8xH_J01_Q4e6Rk/edit?resourcekey=0-mc_vh5UzrzXfU4kO-3tOjA&tab=t.0#heading=h.ank09kkr0tnn), and [2025-03-03b](https://docs.google.com/document/d/1Iut5f2TQBrtBNIduF4vJYOKfw7MbS8xH_J01_Q4e6Rk/edit?resourcekey=0-mc_vh5UzrzXfU4kO-3tOjA&tab=t.0#heading=h.19yyjdek3asm) where this was discussed

Reply:
Added links.